fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LucenePartitioner.java

```diff
@@ -458,4 +483,207 @@ public static long getFrom(@Nonnull LucenePartitionInfoProto.LucenePartitionInfo
+        } else {
+            fetchedRecordsFuture = state.store.fetchIndexRecords(cursor, IndexOrphanBehavior.SKIP).map(FDBIndexedRecord::getStoredRecord).asList();
+        }
+
+        return fetchedRecordsFuture.thenCompose(records -> {
+            if (records.size() > 0) {
+                // the newest record is the one we intend to leave in the current partition; we need it in order to set this partition's new
+                // `from` value.
+                final long newBoundaryTimestamp = getPartitioningTimestampValue(records.get(records.size() - 1));
+
+                // remove the (n + 1)th record from the records to be moved
+                records.remove(records.size() - 1);
+
+                LuceneIndexMaintainer indexMaintainer = (LuceneIndexMaintainer)state.store.getIndexMaintainer(state.index);
+                // shortcut delete docs from current partition
+                // (we do this, instead of calling LuceneIndexMaintainer.update() in order to avoid a chicken-and-egg
+                // situation with the partition metadata keys.
+                records.forEach(r -> {
+                    try {
+                        indexMaintainer.deleteDocument(groupingKey, partitionInfo.getId(), r.getPrimaryKey());
+                    } catch (IOException e) {
+                        throw new RecordCoreException(e);
+                    }
+                });
+
+                // update current partition's meta
+                state.context.ensureActive().clear(partitionMetadataKeyFromTimestamp(groupingKey, getFrom(partitionInfo)));
+                LucenePartitionInfoProto.LucenePartitionInfo.Builder builder = partitionInfo.toBuilder()
+                        .setCount(partitionInfo.getCount() - records.size())
+                        .setFrom(ByteString.copyFrom(Tuple.from(newBoundaryTimestamp).pack()));
+                savePartitionMetadata(groupingKey, builder);
+
+                // value of the "destination" partition's `from` timestamp
+                final long overflowPartitionFromTimestamp = getPartitioningTimestampValue(records.get(0));
+                return findPartitionInfo(groupingKey, overflowPartitionFromTimestamp).thenCompose(previousPartition -> {
+                    if (previousPartition == null || previousPartition.getCount() + records.size() > indexPartitionHighWatermark || previousPartition.getId() == partitionInfo.getId()) {
+                        // create a new "overflow" partition
+                        savePartitionMetadata(groupingKey, newPartitionMetadata(overflowPartitionFromTimestamp, maxPartitionId + 1).toBuilder());
+                    }
+
+                    Iterator<? extends FDBIndexableRecord<Message>> recordIterator = records.iterator();
+                    return AsyncUtil.whileTrue(() -> indexMaintainer.update(null, recordIterator.next())
+                            .thenApply(ignored -> recordIterator.hasNext()));
```

Comment:
It's not a bug, but the layering here feels a little uncomfortable.

Partitions are below the index maintainer: it accepts / returns records and divides work to the appropriate ones.
But the re-partitioner is _above_ the index maintainer. This means (1) you ask the maintainer to delete / update the record and (2) you modify the partition metadata at just the right time and then re-enter records in the same way.
Before seeing this, I would have imagined a partition system below the index maintainer (and above Lucene & FDB) and the re-partitioner as a _sibling_ of the maintainer calling into partitions to remove & add.

I guess I am not actually asking to have anything changed, just capturing my reaction.

Reply:
I had a similar feeling, somewhere else it has to get the index maintainer.
I think the answer is that we need to take some of the `IndexMaintainer` and push it down, so that the `LucenePartitioner` can call `update` on that.